query-planner-js/src/QueryPlan.ts

```diff
@@ -45,6 +51,10 @@ export interface FetchNode {
   outputRewrites?: FetchDataOutputRewrite[];
 }
 
+export interface SubscriptionNode extends Omit<FetchNode, 'kind' | 'hasDefers'> {
```

Comment:
My 2cents, as a consumer of your api it's cleaner to have another specific node for subscription, it's easier and cleaner to integrate in our codebase for example. It's just my opinion from a consumer point of view

Reply:
> it's cleaner to have another specific node for subscription

Just to be extra clear, I'm not suggesting removing `SubscriptionPlanNode`, so there will be specific node for subscription. But I disagree otherwise that, within that already specific node, we need a new type of node to encode the initial fetch. The initial fetch is a normal fetch. It will be send to subgraph completely normally. In particular, why a new type for this primary fetch, but `SubscriptionPlanNode.rest` uses completely normal `FetchNode` within it? What's the rational for that inconsistency?

> it's easier and cleaner to integrate in our codebase for example

I don't mean to sound obtuse, but I have worked on those part of the router and I'd argue the complete opposite.

In fact, my advise on the router side would be to deserialise this as a [`FetchNode`](https://github.com/apollographql/router/blob/dev/apollo-router/src/query_planner/fetch.rs#L64-L93) directly, because you _will_ have to execute it as a normal fetch ultimately (yet again, this _is_ a subgraph fetch, not something else), so if you don't deserialise it as a fetch, you will have to convert to one anyway (or duplicate *tons* of code unnecessarily) and I'm missing how that's cleaner and easier. That is, this doesn't look too bad here only because it's typescript, with the `Omit` and duck-typing; it's going to be more duplication in rust.

But anyway, as far as this typescript code goes, as said above, I would at least insist that this `SubscriptionNode` is not added to `PlanNode` unless there is a justification for that that I don't see. I do think not just reusing `FetchNode` for a fetch is a mistake, but I've obviously make my point enough, so I won't insist on that point anymore.

Reply:
Ah ok, I was only talking about `SubscriptionPlanNode`, my bad. And about implementation in the router, it's not exactly the same than for a fetch node (it's not so far but not really the same). For the different modes we want (websocket and callback) you need to spawn a specific task and do some specific stuffs to communicate between tasks to sync and send to right data to the right client. Happy to show you how it's implemented if you want... But TLDR that's not just a call to a subgraph like for a fetch node, it's a little bit more than that so having a new node is cleaner IMHO